openverse_catalog/dags/providers/provider_api_scripts/phylopic.py

```diff
@@ -13,272 +13,282 @@
-                 DEFAULT_PROCESS_DAYS (7) if undefined.
-    """
-
-    offset = 0
-
-    logger.info("Begin: PhyloPic API requests")
-
-    if date_start == "all":
-        logger.info("Processing all images")
-        param = {"offset": offset}
-
-        image_count = _get_total_images()
-        logger.info(f"Total images: {image_count}")
-
-        while offset <= image_count:
-            _add_data_to_buffer(**param)
-            offset += LIMIT
-            param = {"offset": offset}
-
-    else:
-        if date_end is None:
-            date_end = _compute_date_range(date_start)
-        param = {"date_start": date_start, "date_end": date_end}
-
-        logger.info(f"Processing from {date_start} to {date_end}")
-        _add_data_to_buffer(**param)
-
-    image_store.commit()
-
-    logger.info("Terminated!")
-
-
-def _add_data_to_buffer(**kwargs):
-    endpoint = _create_endpoint_for_IDs(**kwargs)
-    IDs = _get_image_IDs(endpoint)
-    if IDs is None:
-        return
-
-    for id_ in IDs:
-        item_data = _get_object_json(id_)
-        if item_data:
-            _process_item(item_data)
-
-
-def _process_item(item_data: Dict) -> None:
-    details = _get_meta_data(item_data)
-    if details is not None:
-        image_store.add_item(**details)
-
-
-def _get_total_images() -> int:
-    # Get the total number of PhyloPic images
-    total = 0
-    endpoint = "http://phylopic.org/api/a/image/count"
-    result = delayed_requester.get_response_json(endpoint, retries=2)
-
-    if result and result.get("success") is True:
-        total = result.get("result", 0)
-
-    return total
-
-
-def _create_endpoint_for_IDs(**kwargs):
-    limit = LIMIT
-
-    if ((date_start := kwargs.get("date_start")) is not None) and (
-        (date_end := kwargs.get("date_end")) is not None
+class PhylopicDataIngester(ProviderDataIngester):
+    delay = 5
+    host = "http://phylopic.org"
+    endpoint = f"{host}/api/a/image"
+    providers = {"image": prov.PHYLOPIC_DEFAULT_PROVIDER}
+    batch_limit = 25
+    # Default number of days to process if date_end is not defined
+    default_process_days = 7
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        # This is made an instance attribute rather than passed around via query params
+        # because it actually comprises the URL we hit (not the params) depending on
+        # whether we're running a dated DAG or not.
+        self.offset = 0
+
+    @staticmethod
+    def _compute_date_range(date_start: str, days: int = default_process_days) -> str:
+        """
+        Given an ISO formatted date string and a number of days, compute the
+        ISO string that represents the start date plus the days provided.
+        """
+        date_end = date.fromisoformat(date_start) + timedelta(days=days)
+        return date_end.isoformat()
+
+    def get_media_type(self, record: dict) -> str:
+        return constants.IMAGE
+
+    def get_response_json(
```

Comment:
Would it makes sense for `get_next_query_params` to still just return a configuration object that then `get_response_json` converts into a usable endpoint to call `super().get_response_json` with?

I think that might tighten up the implementation and make it seem a little more "normal" and easier to understand. At least I think it would from my perspective, coming into this out of pure curiosity :slightly_smiling_face: 

Another thing that crossed my mind was whether some flag other than empty query params (maybe a sentinel object?) could be used to indicate that the batch operation was complete. It would make the process more explicit for other scripts too as well as allowing this one to avoid the awkward dict copying that requires a whole extra comment to explain it.

Reply:
I've used a totally different method of constructing the endpoints, let me know if this new approach makes more sense!